tap_googleads/streams.py

```diff
@@ -78,90 +87,104 @@ def gaql(self):
                 th.Property("id", th.StringType),
             ),
-        )
+        ),
     ).to_dict()
 
-    # Goal of this stream is to send to children stream a dict of
-    # login-customer-id:customer-id to query for all queries downstream
-    def get_records(self, context: Optional[dict]) -> Iterable[Dict[str, Any]]:
-        """Return a generator of row-type dictionary objects.
+    def post_process(
+        self,
+        row: Record,
+        context: Context | None = None,  # noqa: ARG002
+    ) -> dict | None:
+        """As needed, append or transform raw data to match expected structure.
+
+        Optional. This method gives developers an opportunity to "clean up" the results
+        prior to returning records to the downstream tap - for instance: cleaning,
+        renaming, or appending properties to the raw record result returned from the
+        API.
 
-        Each row emitted should be a dictionary of property names to their values.
+        Developers may also return `None` from this method to filter out
+        invalid or not-applicable records from the stream.
 
         Args:
+            row: Individual record in the stream.
             context: Stream partition or context dictionary.
 
-        Yields:
-            One item per (possibly processed) record in the API.
+        Returns:
+            The resulting record dict, or `None` if the record should be excluded.
+
         """
-        if self.config.get("comma_separated_string_of_customer_ids", False):
-            customer_ids_list = self.config.get("comma_separated_string_of_customer_ids").replace(" ", "").split(",")
-            for i in customer_ids_list:
-                yield {"customerClient": {"id": str(i)}}
-        else:
-            for row in self.request_records(context):
-                row = self.post_process(row, context)
-                # Don't search Manager accounts as we can't query them for everything
-                if row["customerClient"]["manager"] == True:
-                    continue
-                yield row
-
-    def get_child_context(self, record: dict, context: Optional[dict]) -> dict:
+        customer = row["customerClient"]
+
+        if (
+            self.config.get("customer_ids")
+            and customer["id"] not in self.config["customer_ids"]
+        ):
+            return None
+
+        if customer["manager"]:
+            self.logger.warning("%s is a manager, skipping", customer["clientCustomer"])
+            return None
+
+        if customer["status"] != "ENABLED":
+            self.logger.warning(
+                "%s is not enabled, skipping",
+                customer["clientCustomer"],
+            )
+            return None
+
+        return row
+
+    def get_child_context(self, record: dict, context: dict | None) -> dict:  # noqa: ARG002
         """Return a context dictionary for child streams."""
         return {"customer_id": record["customerClient"]["id"]}
```

Comment:
This works on our side. Can you test on yours?

```suggestion
    def get_child_context(self, record: dict, context: dict | None) -> dict:  # noqa: ARG002
        """Return a context dictionary for child streams."""
        customer_id = record["customerClient"]["id"]

        # skip if we've already seen this customer
        if all(
            any(ctx["customer_id"] == customer_id for ctx in cs.partitions or [])
            for cs in self.child_streams
        ):
            return None

        return {"customer_id": customer_id}
```

Reply:
what is the purpose of this?

Reply:
https://github.com/Matatika/tap-googleads/pull/45#discussion_r1766543548

> Yeah, so in our case `AccessibleCustomers` returns both a manager and operating customer account and then `CustomerHierarchyStream` returns the operating customer account for both contexts, meaning that all reports streams are synced twice...

This isn't great for us, and it's possible others have their Google Ads setup in a similar way that this would also be a problem for them. I don't know much at all about user access control in Google Ads, but I don't think we should be syncing streams twice given the same child context regardless. I'm not a massive fan of the above, so if you know of a different way I would be very open to it!

Reply:
that sounds like a different problem than the `customer_id` filter. Is what you're saying that from the `AccessibleCustomer` you get duplicate customer ids?

Reply:
OK, let me try and explain our setup...

Our authenticated principal returns two customers from `AccessibleCustomers` - let's refer to them as `A` and `B`. Since `CustomerHierarchyStream` is a child stream of `AccessibleCustomers`, a hierarchy of customers will be

Reply:
OK, let me try and explain our setup...

Our authenticated principal returns two customers from `AccessibleCustomers` - let's refer to them as `manager` and `operatingCustomer`. Since `CustomerHierarchyStream` is a child stream of `AccessibleCustomers`, a hierarchical list of customers is returned for both `manager` and `operatingCustomer`:
- In the case of `manager`, this includes `manager` itself and `operatingCustomer`
- In the case of `operatingCustomer`, this includes just `operatingCustomer`

We then filter out manager accounts with the logic you introduced, so we are left with two instances of `operatingCustomer` in context (see above `get_child_context`). Since all reports streams are child streams of `CustomerHierarchyStream`, and `operatingCustomer` is set in context twice, all reports streams are run twice.

The above change is to try and prevent duplicate context for `operatingCustomer` by checking if all child streams have already been registered for that particular `customer_id` before setting it.

Reply:
great. thanks for the clarification. I will test it on my end